vm-validator/src/vm_validator.rs

```diff
@@ -42,11 +52,188 @@ pub trait TransactionValidation: Send + Sync + Clone {
+    /// initialized based on configs from the state.
+    fn new(state_view: S) -> Self {
+        let vm = new_vm_for_validation(&state_view);
+        Self {
+            state_view,
+            module_cache: UnsyncModuleCache::empty(),
+            vm,
+        }
+    }
+
+    /// Resets the state view snapshot to the new one. Does not invalidate the module cache, nor
+    /// the VM.
+    fn reset_state_view(&mut self, state_view: S) {
+        self.state_view = state_view;
+    }
+
+    /// Resets the state to the new one, empties module cache, and resets the VM based on the new
+    /// state view snapshot.
+    fn reset_all(&mut self, state_view: S) {
+        self.state_view = state_view;
+        self.module_cache = UnsyncModuleCache::empty();
+        self.vm = new_vm_for_validation(&self.state_view);
+    }
+}
+
+impl<S> WithRuntimeEnvironment for ValidationState<S> {
+    fn runtime_environment(&self) -> &RuntimeEnvironment {
+        self.vm.runtime_environment()
+    }
+}
+
+impl<S: StateView> ModuleCache for ValidationState<S> {
+    type Deserialized = CompiledModule;
+    type Extension = AptosModuleExtension;
+    type Key = ModuleId;
+    type Verified = Module;
+    type Version = usize;
+
+    fn insert_deserialized_module(
+        &self,
+        key: Self::Key,
+        deserialized_code: Self::Deserialized,
+        extension: Arc<Self::Extension>,
+        version: Self::Version,
+    ) -> VMResult<Arc<ModuleCode<Self::Deserialized, Self::Verified, Self::Extension>>> {
+        self.module_cache
+            .insert_deserialized_module(key, deserialized_code, extension, version)
+    }
+
+    fn insert_verified_module(
+        &self,
+        key: Self::Key,
+        verified_code: Self::Verified,
+        extension: Arc<Self::Extension>,
+        version: Self::Version,
+    ) -> VMResult<Arc<ModuleCode<Self::Deserialized, Self::Verified, Self::Extension>>> {
+        self.module_cache
+            .insert_verified_module(key, verified_code, extension, version)
+    }
+
+    fn get_module_or_build_with(
+        &self,
+        key: &Self::Key,
+        builder: &dyn ModuleCodeBuilder<
+            Key = Self::Key,
+            Deserialized = Self::Deserialized,
+            Verified = Self::Verified,
+            Extension = Self::Extension,
+        >,
+    ) -> VMResult<
+        Option<(
+            Arc<ModuleCode<Self::Deserialized, Self::Verified, Self::Extension>>,
+            Self::Version,
+        )>,
+    > {
+        // Get the module that exists in cache.
+        let (module, version) = match self.module_cache.get_module_or_build_with(key, builder)? {
+            None => {
+                return Ok(None);
+            },
+            Some(module_and_version) => module_and_version,
+        };
+
+        // Get the state value that exists in the actual state and compute the hash.
+        let state_value = self
+            .state_view
+            .get_state_value(&StateKey::module_id(key))
```

Comment:
this means we always go to storage, no caching of bytes at all? Even for framework transactions? This could potentially be very expensive? (i.e. imagine 10k tapos transactions)

maybe we should cache the check while the state-view's root hash is the same? (basically on notify commit)

Reply:
There is caching of bytes because here we use `CachedDbStateView` which stores bytes in a locked hashmap, so it's not like we query the DB every time.

Good point about the root.But it will always change, every time a resource is updated? There is no way to check if only code was updated, unless notify commit propagates this info?

Reply:
I was actually going to suggest that we can implement such a mechanism for code - that would be a more optimal way to implement. If perf is okay, I'm fine landing the current way, and maybe adding a note/todo next to the check.

Reply:
Indeed! Added a TODO